### ISSUE TITLE
[2023.11.01_RC3] All print output are redirected into the terminal, if any script tag is using terminal in main thread

### ISSUE THREAD

**Issue by leriomaggio** [pyscript/pyscript#1835]
### Checklist

- [X] I added a descriptive title
- [X] I searched for other issues and couldn't find a solution or duplication
- [X] I already searched in Google and didn't find any good information or help

### What happened?

Release: `[2023.11.01_RC3]`

**Scenario**:

Testing this into an app:
```html
<script type="py">
    print(f"Hello No terminal")
</script>
<script type="py" terminal>
    print("hello in terminal")
</script>
<script type="py">
    print("hello in console")
</script>
```

**Important**: the terminal is instantiated once and in the main thread

**Result**: All the three prints are shown in the termnal regardless. 

Different behaviour occurs if the terminal is spawn in the worker, of course. 

### What browsers are you seeing the problem on? (if applicable)

Chrome

### Additional Context

App tested on PSDC

**Comment by antocuni** [comment]
I think this is the correct/expected behavior (to me, at least), because the stdio redirection can happen only on a per-interpreter basis, and all the tags share the same interpreter.

**Comment by leriomaggio** [comment]
Yeah, but I keep finding this use a bit counter-intuitive, with the ultimate perception that what we have is a sort of (unintentional) side effect.

This, on other hand: 
```html
<script type="py" terminal>
    print(f"Hello No terminal")
</script>
<script type="py" terminal>
    print("hello in terminal")
</script>
<script type="py" terminal>
    print("hello in console")
</script>
```

works very fine, and you would clearly expect to see all condensed in the same terminal (re "we are currently supporting a single terminal per app as in #1834 ) :) 

WDYT?

**Comment by WebReflection** [comment]
that should throw an error per code ... we allow only one terminal per page and we discussed this, and agreed on, that's actually the expectations too ... you get condensed output in the same terminal as long as there is a terminal.

in "classic" the terminal was the default output, in next it's opt in and first come, first works ... nothing to fix here to me?

**Comment by leriomaggio** [comment]
Mmm not sure about the error. I actually see things in exact the opposite direction. 
Allowed you know and document that terminal is indeed unique per app, having multiple `script` tags with `terminal` attribute does not mean "get me a new terminal every time" but rather - "use the terminal, whenever there is any prompt output". 

If there is not, I would like the prompt to be redirected to console only (as per default behaviour, btw!)

In the former case described in the first post, you're changing the defaults and this is bad design, imho.

my2c

**Comment by WebReflection** [comment]
> having multiple script tags with terminal attribute 

this is not desired and not allowed by code. I need to double check that's the case but nobody should advocate for multiple py-script tags in general, as that's never needed or really desired, as there could be only one config per page too, while with workers the terminal is an explicit intent to have an interactive shell and nothing else.

> "use the terminal, whenever there is any prompt output"

that's easy in classic on main only, that cannot work with workers + terminal is not a repl, multiple terminals make no sense by definition or agreement after discussing this.

> If there is not, I would like the prompt to be redirected to console only (as per default behaviour, btw!)

there are two behaviors:
  * main
  * worker

pick your terminal, these are very different too.

> In the former case described in the first post, you're changing the defaults and this is bad design, imho.

if this is a personal attack we've been discussed as a team ad nauseam what we wanted and agreed with the current behavior so ... who is *you* there? :thinking:

**Comment by JeffersGlass** [comment]
> nobody should advocate for multiple py-script tags in general, as that's never needed or really desired,

_(Sidebar)_

I apologize for being late to this party, but I want to push against "one page one pyscript tag" as a goal. As evidenced by the folks here (on GitHub) and in the Discord wanting to use PyScript with React or Django, or in my case using a static site generator like Hugo or MkDocs, the ability to have multiple py-script tags function predictably on the page is important.

_(Sidebar over, here's the relevant part for this issue)_

Personally, I think the fix here for this particular issue is just a docs fix that we add a line to the docs just saying that if _any_ main-thread script tag on the page has a `terminal` attribute, all output from `stdout` will be routed to that terminal.

@leriomaggio I do see what you're saying - it "feels" a bit like the attribute of one script tag is "affecting the behavior" of other tags on the main thread... but as @antocuni says, per-tag choice of whether we're redirecting to terminal or not isn't currently planned. That terminal represents the output of the singular main-thread Interpreter that PyScript provides, so if that output is going to the terminal, all of it is going to the terminal.

Like I say, if it wasn't obvious to you as a user, that's really useful data! I that means we have a documentation hole, but not something I think we should change in the code.

**Comment by WebReflection** [comment]
@JeffersGlass 

> the ability to have multiple py-script tags function predictably on the page is important.

sure thing, and that's understandable, but when it comes to a terminal, which is one per page, that makes less sense as we agreed **for now** there's only one terminal.

I phrased badly my reply but the context here is *multiple terminals* so that *multiple scripts* with a `terminal` tag make little, if none, sense to me. We all agreed about this for the time being, the rest is more like the *dumb terminal* or *cell* Antonio mentioned already a few times, but that's another component.

> Personally, I think the fix here for this particular issue is just a docs fix that we add a line to the docs just saying that if any main-thread script tag on the page has a terminal attribute, all output from stdout will be routed to that terminal.

Agreed, we should improve docs if these are not explicitly stating one terminal per page. I thought that was in already, happy to improve the situation, but I also need to check multiple terminals just early throw as that's code expectation to me, so that's the concrete action as follow up to me. I can try to fix docs too but that's another repository.

**Comment by WebReflection** [comment]
@leriomaggio let's try to understand each other ... if I got it correctly, hence the title of this issue should be changed, what you want is:

  * if multiple scripts with `terminal` attribute are on the same page, on main, no error should exist, they would just default to the very same unique terminal that's shown
  * if multiple scripts with `terminal` also have both *main* and `worker` attribute to target workers, an error should thrown ...
  * we should improve docs around `terminal` so that expectations maybe fail users but are at least well documented

Are these points OK to you?

From a technical PoV though, and just FYI in case you need it, multiple workers terminals are kinda a no-go as the terminal never exits, so that the worker will be "awaited forever" in the queue of scripts that should run ... so that we don't support multiple interactive worker based terminals, but multiple workers also mean multiple pyodide bootstraps per each worker so that, differently from main, these are not ideal for mobile or RAM constrained devices.

If those 3 points are fine to everyone else too, **we** might then go on and provide better ergonomics ... what do others think?

**Comment by fpliger** [comment]
I'd like to set expectations first. This discussion here (a great one btw) is not a release blocker. Independently of what we agree on here, the next release will ship with the current behavior because we intentionally said that's the best path close enough classic that doesn't force us into other weeks.

I personally DO think there's an use case in there and that we should be collecting data and feedback from the community to make sure we have good context and data when we plan this change. That's why releasing now and allowing people to use it and provide feedback is so important.

The architecture now is flexible enough to allow really cool use cases.

**Comment by WebReflection** [comment]
@fpliger agreed but maybe it's worth amending docs if it's not explicit enough how `terminal` behaves? we also just `console.warn` multiple terminals in here (actually it was to indeed allow further iterations instead of throwing out of the blue) and I wonder if we should throw instead ... these are both easy things to amend/fix before the release to me, but you are responsible for the release, so I trust your judgment here.

If nothing, these 2 issues Valerio filed are already some data to consider for the next update, sooner or later.

**Comment by fpliger** [comment]
> @fpliger agreed but maybe it's worth amending docs if it's not explicit enough how terminal behaves?

Yeah. Makes sense. If it's not explicit enough in the docs it creates more confusion to users (whether it's a bug, a feature or an intentional choice for now)

> these are both easy things to amend/fix before the release to me, but you are responsible for the release, so I trust your judgment here.

You mean the change in the docs and, eventually, throwing an error instead of `console.warn` ?

> If nothing, these 2 issues Valerio filed are already some data to consider for the next update, sooner or later.

Yes, exactly! For these features that may evolve in the future, we could even have a sort of discussion open for "proposal for enhancements" or just link to a feedback form (like the [one I was preparing for feedback after we give talks/presentations](https://forms.gle/ohU6SFwmJh9AoC819))

Bottom line is that we start from some intentional choices from educated guesses and the ideal scenario is that we evolve features based on what the community needs.

>  P.S. the other issue is currently this one https://github.com/pyscript/pyscript/issues/1834#issuecomment-1791080568

Yes. +1 on throwing errors do signal the limitations of the current release and what intentionally works and what doesn't

**Comment by WebReflection** [comment]
> You mean the change in the docs and, eventually, throwing an error instead of console.warn ?

yes, these two, I can work on both (at this point tomorrow, but I'll file the throwing MR now if I manage).

**Comment by WebReflection** [comment]
FYI one MR is out https://github.com/pyscript/pyscript/pull/1837 ... this one is about avoiding at all costs multiple py-terminal on a single page so that no confusion should ever land in here about this (yet still open, imho) topic.
